src/main/java/org/openstreetmap/atlas/checks/validation/linear/lines/WaterWayCheck.java

```diff
@@ -0,0 +1,453 @@
+        }
+        return lineItem.getAtlas()
+                .areasCovering(lineItem.asPolyLine().last(), this.waterwaySinkTagFilter::test)
+                .iterator().hasNext();
+    }
+
+    /**
+     * Check if the last location on a line also happens to be connected to a waterway.
+     *
+     * @param line
+     *            The line to check
+     * @return {@code true} if the line ends on a waterway (the last node is part of a waterway)
+     */
+    public boolean doesLineEndOnWaterway(final LineItem line)
+    {
+        final List<LineItem> waterways = new ArrayList<>();
+        line.getAtlas()
+                .lineItemsIntersecting(line.asPolyLine().last().boxAround(Distance.ONE_METER),
+                        this.waterwayTagFilter::test)
+                .forEach(waterways::add);
+
+        waterways.removeIf(line::equals);
+        final Location last = line.asPolyLine().last();
+        return waterways.stream().anyMatch(
+                l -> l.asPolyLine().contains(last) && !last.equals(l.asPolyLine().last()));
+    }
+
+    /**
+     * Check if a location is to the right of a line (e.g., kerbs, cliffs, and oceans)
+     *
+     * @param line
+     *            The line to compare the location to
+     * @param location
+     *            The location
+     * @return {@code true} if the location is to the right of the line. If the location is on the
+     *         line, or to the left of the line, we return {@code false}.
+     */
+    public boolean isRightOf(final PolyLine line, final Location location)
+    {
+        final PolyLine tLine = new PolyLine(location);
+        final Segment closest = line.segments().stream()
+                .min(Comparator.comparingDouble(s -> s.shortestDistanceTo(tLine).asMeters()))
+                .orElse(null);
+        if (closest != null)
+        {
+            final PolyLine testLine = new PolyLine(closest.first(), closest.last(), location);
+            final Angle difference = testLine.headingDifference().orElse(null);
+            if (difference != null && difference.asDegrees() > 0)
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check if a location is inside an atlas bounds
+     *
+     * @param atlas
+     *            The atlas
+     * @param location
+     *            The location to check
+     * @return {@code true} if the atlas contains the location inside its bounds
+     */
+    public boolean isValidEndToCheck(final Atlas atlas, final Location location)
+    {
+        return atlas.bounds().fullyGeometricallyEncloses(location);
+    }
+
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return !isFlagged(object.getOsmIdentifier()) && object instanceof LineItem
+                && this.waterwayTagFilter.test(object);
+    }
+
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
```

Comment:
General question: is it possible to get a waterway that is an offender in multiple ways to this Check? E.g. a waterway that ends in a sink and crosses itself. If so we could append instructions together in one flag instead of returning only one instruction set per flag.

Reply:
It is probably technically possible (so crossing waterways, waterway not properly connected to another waterway, and waterway going in the wrong direction).

I'll look into appending instructions tomorrow.